system-test/bigquery.js

```diff
@@ -505,6 +505,185 @@ describe('BigQuery', function() {
         }
       );
     });
+
+    describe('location', function() {
+      var dataset = bigquery.dataset(generateName('dataset'));
+      var table = dataset.table(generateName('table'));
+      var job;
+
+      var LOCATION = 'asia-northeast1';
+      var QUERY = `SELECT * FROM \`${table.id}\``;
+      var SCHEMA = require('./data/schema.json');
+      var TEST_DATA_FILE = require.resolve('./data/location-test-data.json');
+
+      before(function() {
+        // create a dataset in a certain location will cascade the location
+        // to any jobs created through it
+        return dataset.create({location: LOCATION}).then(function() {
+          return table.create({schema: SCHEMA});
+        });
+      });
+
+      it('should create a load job', function() {
+        return table.createLoadJob(TEST_DATA_FILE).then(function(data) {
+          job = data[0];
+
+          assert.strictEqual(job.location, LOCATION);
+          return job.promise();
+        });
+      });
+
+      describe('job.get', function() {
+        it('should fail to reload if the location is not set', function(done) {
+          var badJob = bigquery.job(job.id);
+
+          badJob.getMetadata(function(err) {
+            assert.strictEqual(err.code, 404);
+            done();
+          });
+        });
+
+        it('should fail to reload if the location is wrong', function(done) {
+          var badJob = bigquery.job(job.id);
+
+          badJob.location = 'US';
+
+          badJob.getMetadata(function(err) {
+            assert.strictEqual(err.code, 404);
+            done();
+          });
+        });
+
+        it('should reload if the location matches', function(done) {
+          var goodJob = bigquery.job(job.id);
+
+          goodJob.location = LOCATION;
+
+          goodJob.getMetadata(function(err) {
+            assert.ifError(err);
+            assert.strictEqual(goodJob.location, LOCATION);
+            done();
+          });
+        });
+      });
+
+      describe('job.cancel', function() {
+        var job;
+
+        before(function() {
+          return dataset.createQueryJob(QUERY).then(function(data) {
+            job = data[0];
+          });
+        });
+
+        it('should fail if the job location is incorrect', function(done) {
+          var badJob = bigquery.job(job.id);
+
+          badJob.location = 'US';
```

Comment:
I think having a dedicated `setLocation()` could be more clear. Prior art: [`file.setEncryptionKey()`](https://github.com/googleapis/nodejs-storage/blob/fb88bd16b25e80962e630a300632d0d15817682c/src/file.js#L1251). To pass it through during instantiation: `bucket.file({ encryptionKey: '...' })`

Do we have other places where we want the user to set a property on a Service/ServiceObject?

Reply:
Not that I'm aware of. I guess I don't see a huge difference between making a dedicated setter vs. allowing the user to configure a public property, however I can definitely get on board with allowing the user to specify the location during instantiation.

Reply:
I like having the setter for consistency now because of Storage, but also for documentation. Having a method with examples and its own docs is much easier to notice. Updating a property doesn’t exactly have the same effect, e.g. in Duplexify you call “setWritable” as opposed to updating a “writable” property.

Reply:
That's true, but I'm not sure if our users will benefit from or ever need to set the language asynchronously. IMO I think passing it in during instantiation is the best way to go.

Reply:
I made a commit to allow passing the location in via constructors. PTAL!

Reply:
So the only way we want the user to set the location is through the constructors, right? Let's remove the getters/setters that allow `{object}.location = '...'` flow. That leaves a backdoor of undocumented behavior that will be confusing to maintain.

Reply:
Okie dokie, I just pushed a commit that removes it.